File: Swift/cyclesort/cyclesort_250000.swift
func cycleSort<T: Comparable>(_ array: inout [T]) {
    let n = array.count
    for cycleStart in 0..<n-1 {
        var item = array[cycleStart]
        var pos = cycleStart
        for i in cycleStart+1..<n {
            if array[i] < item {
                pos += 1
            }
        }
        if pos == cycleStart {
            continue
        }
        while item == array[pos] {
            pos += 1
        }
        let temp = array[pos]
        array[pos] = item
        item = temp
        while pos != cycleStart {
            pos = cycleStart
            for i in cycleStart+1..<n {
                if array[i] < item {
                    pos += 1
                }
            }
            while item == array[pos] {
                pos += 1
            }
            let temp = array[pos]
            array[pos] = item
            item = temp
        }
    }
}


var debug = false

if (debug) {
        print(arr)
}

File: CSHARP/cyclesort/src/cyclesort_250000.cs
using System;

class CycleSort{

    public static bool debug = false;

    public static void cycleSort(int[] array)
    {
        int n = array.Length;

        for (int cycleStart = 0; cycleStart < n - 1; cycleStart++)
        {
            int item = array[cycleStart];
            int pos = cycleStart;

            for (int i = cycleStart + 1; i < n; i++)
            {
                if (array[i] < item)
                {
                    pos++;
                }
            }

            if (pos == cycleStart)
            {
                continue;
            }

            while (item == array[pos])
            {
                pos++;
            }

            int temp = array[pos];
            array[pos] = item;
            item = temp;

            while (pos != cycleStart)
            {
                pos = cycleStart;

                for (int i = cycleStart + 1; i < n; i++)
                {
                    if (array[i] < item)
                    {
                        pos++;
                    }
                }

                while (item == array[pos])
                {
                    pos++;
                }

                temp = array[pos];
                array[pos] = item;
                item = temp;
            }
        }
    }

    public static void Main()
    {
        
        cycleSort(ARRAY);
        if(debug){
            for(int i = 0; i < ARRAY.Length; i++){
                    Console.Write(ARRAY[i] + " ");
            }
        }
              
    }
}